### PR TITLE
Ensure that CA initialization does not block leader election.

### LINKED ISSUE #9498: Preventing Infinite Rounds of Leader Establishment
### Problem

Currently the process of establishing a leader for set of consul servers involves a few things.

1. Election of a leader via Raft.
2. Initializing ACLs
3. Setting up KV deletion management timers. *We keep deleted KV entries around a while longer to prevent the blocking query index from going backwards*
4. Initializing session timers.
5. Initializing namespaces. *Enterprise only*
6. Bootstrap Config Entries. *Config entries put into the Consul configuration will get inserted if they don’t already exist*
7. Initializing and starting autopilot.
8. Initializing the Connect CA
9. Starting config entry replication
10. Starting Connect leader routines. *There are a few of these which perform activities such as monitoring of the primary dc’s roots and refreshing the secondary dc’s intermediate certificate*

The failure of many of the following steps can halt leader establishment.

* ACL initialization
* Session timer initialization
* Namespace initialization
* Config entry bootstrapping
* Connect CA initialization

Some of these we can have a reasonable expectation that there failure would indicate something intrinsically wrong with a single node.

* Session timer initialization - For this one it can only fail if we fail to lookup sessions from memdb or if we fail to reset timers associated with the session. Both cases should generally be impossible in the absence of some sort of memory corruption.
* Namespace initialization - Here we simply insert the `default` namespace definition if it does not exist. This would require memdb or raft to be inoperable for it to fail.
* ACL initialization - This “step” actually performs a number of operations such as inserting the master token, anonymous token and the global management policy. Additionally there is a step performed in secondary datacenters to cleanup tokens affected by a security issue fixed back in the 1.7.x timeframe. All of these operations basically read from memdb to see if data is there and if not inserts it via Raft.

The two that can halt leader establishment but cannot necessarily imply an issue with the node itself are:

* Config entry bootstrapping
* Connect CA initialization

Config entry bootstrapping is similar to namespace initialization in that all its doing is reading from memdb and conditionally inserting new data through raft. However where it differs is that the data being inserted is wholly controlled by the user writing the configuration. Config entries also have dependencies between each other and must be in the proper order within the configuration.  Therefore if the configuration is syntactically valid but the dependencies and ordering between the config entries isn’t correct, it will prevent establishing a leader. Assuming that all the servers have the same configuration then it would be impossible to elect a leader with the only remediation being to fix the configurations and restart the servers. This is bad but at least its fixable by an operator such as the HashiCorp team operating Consul for HCP/HCS.

Connect CA initialization involves many steps but they generally fall into the categories of ca provider configuration, certificate generation or memdb state management. The memdb state management routines can fail but they would indicate either a Raft/memdb failure similar to namespace initialization or acl initialization. The ca provider configuration involves interfaces for various CA provider implementations. As its an interface extension point this could fail for reasons related to an issue on the local node or could involve issues that would affect all nodes and are outside our control. Similarly the certificate generation failure modes do involve things outside our control. If you are using the Vault or AWS PCA CA providers and the external systems are not available then the process of generating certificates would fail and halt leader establishment. One big problem with this is that in order to fix the problem related to how the Consul servers might contact the CA provider it requires a leader to make the changes.

### Suggested Changes
Connect CA initialization should not be able to halt leader establishment. As this process can have dependencies on external systems (or even just other Consul datacenters), we need to prevent network connectivity issues or failures in those remote systems from preventing Consul leader establishment. For some added context there is issue #7529 which is a bug report regarding the inability of a set of servers to establish a leader due to region configuration issues.

Whether config entry bootstrapping should be made to not halt the process is debatable. At least in this case operators who provide access to others cannot have those others cause issues in the cluster. Additionally and maybe more importantly, if there is an issue the operator can simply fix the config. 

### Implementation Notes

It will not be sufficient to just suppress the errors that could happen during Connect CA initialization. Instead we will probably need to start go routines to periodically retry that initialization.

### PR DESCRIPTION
**Pull request by mkeeler**
Fixes #9498

After fixing that bug I uncovered a couple more:

1. Fix an issue where we might try to cross sign a cert when we never had a valid root.
2. Fix a potential issue where reconfiguring the CA could cause either the Vault or AWS PCA CA providers to delete resources that are still required by the new incarnation of the CA.

The main fix was to remove the call to `InitializeCA` from within `establishLeadership`. Beyond that I made the `CAManager` have a `Start` method whose responsibility is to attempt to init the CA and if it fails spawn some background routines to ensure that will happen. Once CA initialization is complete the other background routines for performing cert renewal and primary dc roots watches are started.

### CODE REVIEW

File: agent/consul/leader_connect_test.go [hunk 1]
```diff
@@ -1204,3 +1204,102 @@ func TestLeader_retryLoopBackoffHandleSuccess(t *testing.T) {
...
+	defer s1.Shutdown()
+
+	waitForLeaderEstablishment(t, s1)
+
+	rootsList, activeRoot, err := getTestRoots(s1, "dc1")
+	require.NoError(t, err)
+	require.Empty(t, rootsList.Roots)
+	require.Nil(t, activeRoot)
+
+	newConfig := &structs.CAConfiguration{
```
Comment (banks) [inline agent/consul/leader_connect_test.go:1236]: nit: I was about to leave a note about needing to test this case too based on the test names alone as they only talk about "not preventing leader establishment".

It's probably enough though just to put a comment in something like:

```go
// Now leader is up, verify we can actually reconfigure away from the bad configuration
```

What do you think?

**Review by banks — APPROVED** [review]
This is looking good overall.

The one thing I didn't see that I'm not sure about UX wise is what happens if you try to hit the CA endpoints after leader establishment but before the CA is configured properly?

It looks like you'd see these errors: https://github.com/hashicorp/consul/blob/7b7589eea8b8f40f304d9487a9676af978081602/agent/consul/server_connect.go#L135-L140

Which is better than a crash but would it be simple to be more friendly and actually check the state somehow and return a different error when you hit the case that the CA was unable to initialize on election and is retrying in the background?

File: agent/connect/ca/provider_aws_test.go [hunk 2]
```diff
@@ -235,6 +237,143 @@ func TestAWSNoCrossSigning(t *testing.T) {
...
+		// changing the ARN should cause the other CA to be deleted
+		p2ConfAltARN := testProviderConfigPrimary(t, map[string]interface{}{
+			"ExistingARN": "doesnt-need-to-be-real",
+		})
+		require.NoError(t, p2.Cleanup(false, p2ConfAltARN.RawConfig))
+
+		requirePCADeleted(t, p1)
+	})
+
+	t.Run("arn-change", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/connect/ca/provider_aws_test.go:347]: This was a copy-paste error. There were two subtests with the same name. This second one should have been renamed.

File: agent/consul/leader_connect_ca.go [hunk 3]
```diff
@@ -234,21 +240,83 @@ func (c *CAManager) setCAProvider(newProvider ca.Provider, root *structs.CARoot)
...
+	if oldState == CAStateInitialized {
+		return nil
+	}
 	if err != nil {
 		return err
 	}
-	defer c.setState(CAStateReady, false)
+
+	defer func() {
+		if reterr == nil {
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/consul/leader_connect_ca.go:314]: I added a comment and a link to a golang.org blog post on the matter: https://blog.golang.org/defer-panic-and-recover

File: agent/consul/leader_connect_ca.go [hunk 4]
```diff
@@ -22,7 +22,7 @@ type CAState string
 const (
 	CAStateUninitialized     CAState = "UNINITIALIZED"
 	CAStateInitializing              = "INITIALIZING"
-	CAStateReady                     = "READY"
+	CAStateInitialized               = "INITIALIZED"
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/consul/leader_connect_ca.go:25]: Yes. For good measure I also went ahead and unexported the type and all the constant names as well. Currently this code lives in the `agent/consul` package so plenty of things still have the potential to use them. However vscode/gopls tell me that the type isn't reference anywhere outside of this file.

File: agent/connect/ca/provider.go [hunk 5]
```diff
@@ -162,7 +162,7 @@ type Provider interface {
 	// Cleanup performs any necessary cleanup that should happen when the provider
 	// is shut down permanently, such as removing a temporary PKI backend in Vault
 	// created for an intermediate CA.
-	Cleanup() error
+	Cleanup(providerTypeChange bool, config map[string]interface{}) error
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/connect/ca/provider.go]: I changed it to be referred to as the `otherConfig`. Depending on success vs failure we might call Cleanup on a new provider instance or the old one. Basically whenever there are two provider instances during a reconfiguration we will call `Cleanup` on one of them passing in the other providers config. 

That way the provider can check it against its own config to determine what should be deleted or not.

**Comment by mkeeler** [comment]
> The one thing I didn't see that I'm not sure about UX wise is what happens if you try to hit the CA endpoints after leader establishment but before the CA is configured properly?
> 
> It looks like you'd see these errors:
> 
> https://github.com/hashicorp/consul/blob/7b7589eea8b8f40f304d9487a9676af978081602/agent/consul/server_connect.go#L135-L140
> 
> Which is better than a crash but would it be simple to be more friendly and actually check the state somehow and return a different error when you hit the case that the CA was unable to initialize on election and is retrying in the background?

That error is what you see if you try to sign a leaf/intermediate before the CA is fully initialized. Other endpoints such as v1/connect/ca/roots returns a valid response that is mostly empty (Active Root ID field is `""` and the `RootCerts` field is empty).

Maybe it would be enough to just change that error message for when the root is nil to something like: `CA is still being initialized and is unable to sign certificates`

Something that makes it sound like some state we are in and have accounted for rather than some random unexpected state.

I thought about using the internal state of the CA Manager to tell consumers that the CA cannot be used yet. But really thats the same as having a non-nil root (for the most part, there are some "bugs" with that documented in #9573 and should be fixed separately)

**Comment by mkeeler** [comment]
@banks I ended up just improving both the error messages in server_connect.go a little bit to make it more descriptive and sound less like we have no idea why we could be in those states.

I also found one potential issue which is that it was possible for CA reconfiguration to take place without ever successfully initializing the CA configuration including the trust domain. I added a call into UpdateConfiguration of the CA manager to initialize the config if it hadn't already been done.

File: agent/connect/ca/provider_aws.go [hunk 6]
```diff
@@ -232,7 +232,7 @@ func (a *AWSProvider) ensureCA() error {
 	}
 
 	a.rootPEM = certPEM
-	return nil
+	return nil //
```
Comment (banks) [inline agent/connect/ca/provider_aws.go:235]: nit: is this an empty comment?

**Review by banks — APPROVED** [review]
> I added a call into UpdateConfiguration of the CA manager to initialize the config if it hadn't already been done.

Good catch yeah that would have potentially been unpleasant.

Stil LGTM

File: agent/connect/ca/provider_aws.go (hunk 6 above)
Comment (mkeeler) [reply agent/connect/ca/provider_aws.go:235]: That was definitely accidental. Removing now.

**Comment by mkeeler** [comment]
The failed CI job is due to not rebasing to pull in some recent CI changes to circle.

**Pull request merged**